aten/src/ATen/BatchedFallback.cpp

```diff
@@ -0,0 +1,154 @@
+  SmallVector<Tensor,kVmapTransformStaticInputSize> batched_tensor_inputs;
+  VmapDimVector batched_tensor_inputs_position;
+  for (int64_t idx = 0; idx < arguments.size(); ++idx) {
+    const auto& ivalue = arguments[idx];
+    if (!ivalue.isTensor()) {
+      continue;
+    }
+    const auto& tensor = ivalue.toTensor();
+    if (!tensor.defined()) {
+      continue;
+    }
+    const auto* batched = maybeGetBatched(tensor);
+    if (!batched) {
+      continue;
+    }
+    batched_tensor_inputs.push_back(tensor);
+    batched_tensor_inputs_position.push_back(idx);
+  }
+  TORCH_INTERNAL_ASSERT(batched_tensor_inputs.size() > 0);
+
+  // MultiBatchVmapTransform the BatchedTensor arguments. This returns
+  // VmapPhysicalViews that contain all of the batch dimensions.
+  const auto input_physical_views = MultiBatchVmapTransform::logicalToPhysical(
+      batched_tensor_inputs);
+
+  // Compute the total number of batches
+  auto num_batch_dims = input_physical_views.front().numBatchDims();
+  auto some_sizes = input_physical_views.front().tensor().sizes();
+  auto batch_sizes = ArrayRef<int64_t>(some_sizes.begin(), some_sizes.begin() + num_batch_dims);
+  auto num_batches = std::accumulate(
+      batch_sizes.begin(),
+      batch_sizes.end(),
+      1,
+      std::multiplies<int64_t>());
+
+  // Strategy: For each batch, we are going to push slices (where applicable)
+  // of the arguments onto `stack`, call `op`, and store the result in
+  // `output_shards`.
+  std::vector<Tensor> output_shards;
+  output_shards.reserve(num_batches);
+
+  for (int64_t linear_idx = 0; linear_idx < num_batches; ++linear_idx) {
+    auto index = computeIndex(linear_idx, batch_sizes);
+    auto batched_tensor_inputs_pos_iter = batched_tensor_inputs_position.begin();
+    auto input_physical_views_iter = input_physical_views.begin();
+    for (int64_t arg_idx = 0; arg_idx < num_arguments; ++arg_idx) {
+      // The accounting for this is tricky. Assume `op` has 4 args, A, B, C, D.
+      // and `arg_idx = 2`. Every iteration of this loop pushes one element
+      // onto the stack, so we've already pushed two things (call them A' and B').
+      // At this point, the stack should look like the following:
+      // [..., A, B, C, D, A', B'] (top of stack here)
+      // To get at C (argument number 2), we have to peek `num_arguments`
+      // into the stack from the top
+      const auto& argument = torch::jit::peek(stack, 0, num_arguments);
+      if (arg_idx != *batched_tensor_inputs_pos_iter) {
+        // argument isn't a BatchedTensor
+        torch::jit::push(stack, argument);
+        continue;
+      }
+      // argument is a BatchedTensor
+      const auto& physical_view_for_argument = *input_physical_views_iter;
+      torch::jit::push(stack, physical_view_for_argument.tensor().index(index));
+      batched_tensor_inputs_pos_iter++;
+      input_physical_views_iter++;
+    }
+
+    op.callBoxed(stack);
+
+    // We assume there is a single tensor return
```

Comment:
Hmmmmmmmm

Reply:
This is probably not true in general

Reply:
Yeah, that'll get fixed in the future :) (probably one of the next three diffs)

Reply:
I tried implementing support for multiple returns but I ran into https://github.com/pytorch/pytorch/issues/42378 (there's a minimal repro in the example). Should an operator with multiple returns push all of them onto the stack?